src/database-controller/src/write-merger/handler.js

```diff
@@ -0,0 +1,176 @@
+          await databaseModel.Framework.create(record);
+        } else {
+          throw createError(404, `Cannot find framework ${frameworkName}.`);
+        }
+      } else {
+        // Database has the corresponding framework.
+        const oldSnapshot = new Snapshot(oldFramework.snapshot);
+        const internalUpdate = {};
+        if (oldSnapshot.getGeneration() === snapshot.getGeneration()) {
+          // if framework request is equal, mark requestSynced = true
+          logger.info(`The request of framework ${frameworkName} is synced.`);
+          internalUpdate.requestSynced = true;
+        } else {
+          // if framework request is not equal,
+          // should use framework request in db as ground truth
+          internalUpdate.requestSynced = false;
+        }
+        // use request in database
+        snapshot.overrideRequest(oldSnapshot);
+        if (req.params.eventType === 'DELETED') {
+          // if event is DELETED, mark apiServerDeleted = true
+          internalUpdate.apiServerDeleted = true;
+        }
+        await databaseModel.Framework.update(
+          _.assign(snapshot.getStatusUpdate(), internalUpdate),
+          { where: { name: frameworkName } },
+        );
+      }
+    });
+    res.status(200).json({ message: 'ok' });
+  } catch (err) {
+    return next(err);
+  }
+}
+
+async function receiveFrameworkRequest(req, res, next) {
+  try {
+    const {
+      frameworkRequest,
+      submissionTime,
+      configSecretDef,
+      priorityClassDef,
+      dockerSecretDef,
+    } = req.body;
+    const frameworkName = _.get(frameworkRequest, 'metadata.name');
+    if (!frameworkName) {
+      return next(createError(400, 'Cannot find framework name.'));
+    }
+    const [needSynchronize, snapshot, addOns] = await lock.acquire(
+      frameworkName,
+      async () => {
+        const oldFramework = await databaseModel.Framework.findOne({
+          attributes: ['snapshot'],
+          where: { name: frameworkName },
+        });
+        const snapshot = new Snapshot(frameworkRequest);
+        const addOns = new AddOns(
+          configSecretDef,
+          priorityClassDef,
+          dockerSecretDef,
+        );
+        if (!oldFramework) {
+          // create new record in db
+          // including all add-ons and submissionTime
+          // set requestGeneration = 1
+          snapshot.setGeneration(1);
+          const record = _.assign(
+            {},
+            snapshot.getAllUpdate(),
+            addOns.getUpdate(),
+          );
+          record.submissionTime = new Date(submissionTime);
+          await databaseModel.Framework.create(record);
+          return [true, snapshot, addOns];
+        } else {
+          // update record in db
+          const oldSnapshot = new Snapshot(oldFramework.snapshot);
+          // compare framework request (omit requestGeneration)
+          if (
+            _.isEqual(snapshot.getRequest(true), oldSnapshot.getRequest(true))
```

Comment:
>isEqual [](start = 14, length = 7)

Is it deep and semantic equal?

Reply:
https://lodash.com/docs/4.17.15#isEqual

It is deep equal. What does semantic equal refer to ?

Reply:
It refers to not only just string compare, i.e. set {1, 2} = set {2, 1}
The lib looks ok